lib/bridge/IrcHandler.js

```diff
@@ -379,47 +379,85 @@ IrcHandler.prototype.onPart = Promise.coroutine(function*(req, server, leavingUs
 
-        var promises = matrixRooms.map((room) => {
-            req.log.info((enabled ? "Locking room %s" :
-                "Reverting %s back to default join_rule"),
-                room.getId()
-            );
-            if (enabled) {
-                return this._setMatrixRoomAsInviteOnly(room, true);
-            }
-            // don't "unlock"; the room may have been invite
-            // only from the beginning.
-            enabled = server.getJoinRule() === "invite";
-            return this._setMatrixRoomAsInviteOnly(room, enabled);
-        });
+    // For k and i, we currently want to flip the join_rules to be
+    // 'invite' to prevent new people who are not in the room from
+    // joining.
 
-        yield Promise.all(promises);
+    // For s, we just want to control the room directory visibility
+    // accordingly. (+s = 'private', -s = 'public')
+
+    // TODO: Add support for specifying the correct 'keyword' and
+    // support for sending INVITEs for virtual IRC users.
+    let matrixRooms = yield this.ircBridge.getStore().getMatrixRoomsForChannel(server, channel);
+    if (matrixRooms.length === 0) {
+        req.log.info("No mapped matrix rooms for IRC channel %s", channel);
+        return;
     }
+
+    var promises = matrixRooms.map((room) => {
+        switch (mode) {
+            case "s":
+                let cli = this.ircBridge.getAppServiceBridge().getClientFactory().getClientAs();
+                return cli.setRoomDirectoryVisibility(room.getId(), enabled ? 'private' : 'public');
+            case "k":
+            case "i":
+                req.log.info((enabled ? "Locking room %s" :
+                    "Reverting %s back to default join_rule"),
+                    room.getId()
+                );
+                if (enabled) {
+                    return this._setMatrixRoomAsInviteOnly(room, true);
+                }
+                // don't "unlock"; the room may have been invite
+                // only from the beginning.
+                enabled = server.getJoinRule() === "invite";
+                return this._setMatrixRoomAsInviteOnly(room, enabled);
+            default:
+                // Not reachable, but warn anyway in case of future additions
+                log.warn(`onMode: Unhandled channel mode ${mode}`);
+                return Promise.resolve();
+        }
+    });
+
+    yield Promise.all(promises);
+});
+
+/**
+ * Called when channel mode information is received
+ * @param {Request} req The metadata request
+ * @param {IrcServer} server : The sending IRC server.
+ * @param {string} channel The channel that has the given mode.
+ * @param {string} mode The mode that the channel is in, e.g. +sabcdef
+ * @return {Promise} which is resolved/rejected when the request finishes.
+ */
+IrcHandler.prototype.onModeIs = Promise.coroutine(function*(req, server, channel, mode) {
+    req.log.info(`onModeIs for ${channel} = ${mode}.`);
+
+    mode = mode.slice(1); // Remove '+'
```

Comment:
For robustness I'd consider instead trimming a leading `+` if one exists, using a regexp. I could imagine some silly ircd somewhere not bothering with a leading `+` and so we'll ignore the first mode flag.

Maybe

    if (mode.match(/^\+/)) mode = mode.slice(1);